lib/Runtime/PlatformAgnostic/Platform/Common/Intl.cpp

```diff
@@ -0,0 +1,487 @@
+#endif
+
+#include "Common/MathUtil.h"
+#include "Core/AllocSizeMath.h"
+
+#include "Intl.h"
+#include "IPlatformAgnosticResource.h"
+
+#define U_STATIC_IMPLEMENTATION
+#define U_SHOW_CPLUSPLUS_API 1
+#pragma warning(push)
+#pragma warning(disable:4995) // deprecation warning
+#include <unicode/uloc.h>
+#include <unicode/numfmt.h>
+#include <unicode/enumset.h>
+#include <unicode/decimfmt.h>
+#pragma warning(pop)
+
+namespace PlatformAgnostic
+{
+namespace Intl
+{
+    using namespace PlatformAgnostic::Resource;
+
+    bool IsWellFormedLanguageTag(_In_z_ const char16 *languageTag, _In_ const charcount_t cch)
+    {
+        bool success = false;
+        UErrorCode error = UErrorCode::U_ZERO_ERROR;
+        char icuLocaleId[ULOC_FULLNAME_CAPACITY] = { 0 };
+        char icuLangTag[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int32_t parsedLength = 0;
+        int32_t forLangTagResultLength = 0;
+        int32_t toLangTagResultLength = 0;
+        size_t inputLangTagUtf8SizeActual = 0;
+
+        // Allocate memory for the UTF8 output buffer. Need 3 bytes for each (code point + null) to satisfy SAL.
+        const size_t inputLangTagUtf8SizeAllocated = AllocSizeMath::Mul(AllocSizeMath::Add(cch, 1), 3);
+        unsigned char *inputLangTagUtf8 = new unsigned char[inputLangTagUtf8SizeAllocated];
+        StringBufferAutoPtr<unsigned char> guard(inputLangTagUtf8);
+
+        if (!inputLangTagUtf8)
+        {
+            AssertOrFailFastMsg(false, "OOM: failed to allocate inputLangTagUtf8");
+            goto LReturn;
+        }
+
+        // explicitly zero the allocated array
+        for (size_t index = 0; index < inputLangTagUtf8SizeAllocated; ++index)
+        {
+            inputLangTagUtf8[index] = 0;
+        }
+
+        inputLangTagUtf8SizeActual = utf8::EncodeIntoAndNullTerminate(inputLangTagUtf8, languageTag, cch);
+
+        // Convert input language tag to a locale ID for use in uloc_toLanguageTag API.
+        forLangTagResultLength = uloc_forLanguageTag(reinterpret_cast<const char *>(inputLangTagUtf8),
+            icuLocaleId, ULOC_FULLNAME_CAPACITY, &parsedLength, &error);
+        success = (forLangTagResultLength > 0) && (parsedLength > 0) &&
+            U_SUCCESS(error) && ((size_t)parsedLength == inputLangTagUtf8SizeActual);
+        if (!success)
+        {
+            goto LReturn;
+        }
+
+        toLangTagResultLength = uloc_toLanguageTag(icuLocaleId, icuLangTag, ULOC_FULLNAME_CAPACITY, TRUE, &error);
+        success = toLangTagResultLength && U_SUCCESS(error);
+        if (!success)
+        {
+            if (error == UErrorCode::U_ILLEGAL_ARGUMENT_ERROR)
+            {
+                AssertMsg(false, "uloc_toLanguageTag: error U_ILLEGAL_ARGUMENT_ERROR");
+            }
+            else
+            {
+                AssertMsg(false, "uloc_toLanguageTag: unexpected error (besides U_ILLEGAL_ARGUMENT_ERROR)");
+            }
+
+            goto LReturn;
+        }
+
+    LReturn:
+        return success;
```

Comment:
I am assuming this used to have a `delete[]` that got removed with the addition of StringBufferAutoPtr, but it may be worth it to remove this label/goto path as a result.

Reply:
Yes, you're correct. Now no explicit cleanup is needed so we can change back to multiple returns instead of using goto/label. #Pending